### PR TITLE
Fix overwrite of ceph-fuse's log and sock file

### LINKED ISSUE #66664: cephfs volume doesn't support client quota
**Is this a BUG REPORT or FEATURE REQUEST?**:
/kind bug

**What happened**:
Cephfs volume with ceph-fuse of ```jewel``` and earlier version does not support capicity quota, which means that the pod using the cephfs volume can use as many space as the cephfs provides, instead of the quota's limitation.

**What you expected to happen**:
Pods mounting cephfs volumes should be rejected to write any more when a quota limit is reached. To be exactly, kubelet should use ```--client-quota``` option when mounting cephfs volume to pod with ceph-fuse of ```jewel``` and earlier version.

**How to reproduce it (as minimally and precisely as possible)**:
Create a cephfs volume of 100MiB capacity by pvc, then mount the volume to a pod on a kubelet node with ceph-fuse of ```jewel``` and earlier version, run df in the container, the 'Available size' will not be 100MiB, but the total capicity of cephfs.

**Anything else we need to know?**:
ceph-fuse of ```luminous``` and later version does not support ```--client-quota``` option, as far as I know. The quota takes effect automaticlly. So we have to be careful not to add ```--client-quota``` option on ceph-fuse who does not recognize it.

**Environment**:
- Kubernetes version (use `kubectl version`): v1.11.1
- Cloud provider or hardware configuration:
- OS (e.g. from /etc/os-release): CentOS Linux release 7.2.1511 (Core) 
- Kernel (e.g. `uname -a`): Linux 4.17.0-1.el7.elrepo.x86_64
- Install tools: kubespray
- Others: ceph-fuse 10.2.5

### PR DESCRIPTION
**Pull request by choury**
**What this PR does / why we need it**:
The default log file of ceph-fuse is  /var/log/ceph-client-{id}.log, it will be overwrite if mount twice.
Same as sock file.
Fix #66664

**Special notes for your reviewer**:

**Release note**:

```release-note
Fix overwrite of ceph-fuse's log and sock file
```

### CODE REVIEW

**Comment by rootfs** [comment]
cc @cofyc @zhangxiaoyu-zidif

File: pkg/volume/cephfs/cephfs.go [hunk 1]
```diff
@@ -335,8 +339,8 @@ func (cephfsMounter *cephfsMounter) checkFuseMount() bool {
 	execute := cephfsMounter.plugin.host.GetExec(cephfsMounter.plugin.GetPluginName())
 	switch runtime.GOOS {
 	case "linux":
-		if _, err := execute.Run("/usr/bin/test", "-x", "/sbin/mount.fuse.ceph"); err == nil {
-			glog.V(4).Info("/sbin/mount.fuse.ceph exists, it should be fuse mount.")
+		if _, err := execute.Run("/usr/bin/ceph-fuse", "--version"); err == nil {
```
Comment (rootfs) [inline pkg/volume/cephfs/cephfs.go]: @zhangxiaoyu-zidif i think this is what in the first ceph-fuse, what happened since then?

File: pkg/volume/cephfs/cephfs.go [hunk 2]
```diff
@@ -407,6 +411,12 @@ func (cephfsVolume *cephfs) execFuseMount(mountpoint string) error {
 	mountArgs = append(mountArgs, cephfsVolume.path)
 	mountArgs = append(mountArgs, "--id")
 	mountArgs = append(mountArgs, cephfsVolume.id)
+	mountArgs = append(mountArgs, "--client_reconnect_stale")
```
Comment (rootfs) [inline pkg/volume/cephfs/cephfs.go]: @batrick are you comfortable with these options?

File: pkg/volume/cephfs/cephfs.go [hunk 3]
```diff
@@ -288,12 +289,15 @@ func (cephfsVolume *cephfs) GetPath() string {
 	return cephfsVolume.plugin.host.GetPodVolumeDir(cephfsVolume.podUID, utilstrings.EscapeQualifiedNameForDisk(name), cephfsVolume.volName)
 }
 
-// GetKeyringPath creates cephfuse keyring path
-func (cephfsVolume *cephfs) GetKeyringPath() string {
+// GetKeyringPath creates cephfuse keyring file name
+func (cephfsVolume *cephfs) GetKeyringFileName() string {
```
Comment (rootfs) [inline pkg/volume/cephfs/cephfs.go]: getKeyringFileName instead?

File: pkg/volume/cephfs/cephfs.go [hunk 4]
```diff
@@ -288,12 +289,15 @@ func (cephfsVolume *cephfs) GetPath() string {
...
+	return cephfsVolume.id + ".keyring"
+}
+
+// GetConfigPath creates cephfuse config path
+func (cephfsVolume *cephfs) GetConfigPath() string {
```
Comment (rootfs) [inline pkg/volume/cephfs/cephfs.go]: getConfigPath

File: pkg/volume/cephfs/cephfs.go (hunk 1 above)
Comment (choury) [reply pkg/volume/cephfs/cephfs.go]: In some distro, the path is /usr/sbin/mount.fuse.ceph. I think if we use ceph-fuse, we should just check it.

File: pkg/volume/cephfs/cephfs.go [hunk 5]
```diff
@@ -242,7 +243,7 @@ func (cephfsVolume *cephfsMounter) SetUpAt(dir string, fsGroup *int64) error {
 		glog.V(4).Info("CephFS fuse mount.")
 		err = cephfsVolume.execFuseMount(dir)
 		// cleanup no matter if fuse mount fail.
-		keyringPath := cephfsVolume.GetKeyringPath()
+		keyringPath := path.Join(cephfsVolume.getConfigPath(), cephfsVolume.getKeyringFileName())
```
Comment (cofyc) [inline pkg/volume/cephfs/cephfs.go]: use `filepath.Join` instead of `path.Join`

File: pkg/volume/cephfs/cephfs.go (hunk 1 above)
Comment (cofyc) [reply pkg/volume/cephfs/cephfs.go]: Did some searching, found that the first version uses `/sbin/mount.fuse.ceph`  too, see https://github.com/kubernetes/kubernetes/pull/55866/files#diff-021042d70c9aac8994638a8d0827b1e0R330.

It's better to check `/usr/bin/ceph-fuse` because we execute `/usr/bin/ceph-fuse` command instead of `/sbin/mount.fuse.ceph` in `execFuseMount` function. It's more consistent.

If we require `ceph-fuse` must be at `/usr/bin/ceph-fuse`, I suggest to use absolute path `/usr/bin/ceph-fuse` instead of `ceph-fuse` in [`execFuseMount`](https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/cephfs/cephfs.go#L412).

File: pkg/volume/cephfs/cephfs.go [hunk 6]
```diff
@@ -407,6 +411,12 @@ func (cephfsVolume *cephfs) execFuseMount(mountpoint string) error {
 	mountArgs = append(mountArgs, cephfsVolume.path)
 	mountArgs = append(mountArgs, "--id")
 	mountArgs = append(mountArgs, cephfsVolume.id)
+	mountArgs = append(mountArgs, "--client_reconnect_stale")
+	mountArgs = append(mountArgs, "--client-quota")
+	mountArgs = append(mountArgs, "--admin_socket")
+	mountArgs = append(mountArgs, filepath.Join(configPath, fmt.Sprintf("ceph-fuse.%s.asok", cephfsVolume.id)))
```
Comment (cofyc) [inline pkg/volume/cephfs/cephfs.go]: type: `asok` -> `sock`

File: pkg/volume/cephfs/cephfs.go [hunk 7]
```diff
@@ -288,12 +289,15 @@ func (cephfsVolume *cephfs) GetPath() string {
 	return cephfsVolume.plugin.host.GetPodVolumeDir(cephfsVolume.podUID, utilstrings.EscapeQualifiedNameForDisk(name), cephfsVolume.volName)
 }
 
-// GetKeyringPath creates cephfuse keyring path
-func (cephfsVolume *cephfs) GetKeyringPath() string {
+// getKeyringPath creates cephfuse keyring file name
+func (cephfsVolume *cephfs) getKeyringFileName() string {
+	return cephfsVolume.id + ".keyring"
```
Comment (cofyc) [inline pkg/volume/cephfs/cephfs.go]: IIUC, this `id` is not unique, perhaps it's better to use some unique string? cc @rootfs

File: pkg/volume/cephfs/cephfs.go [hunk 8]
```diff
@@ -407,6 +411,12 @@ func (cephfsVolume *cephfs) execFuseMount(mountpoint string) error {
 	mountArgs = append(mountArgs, cephfsVolume.path)
 	mountArgs = append(mountArgs, "--id")
 	mountArgs = append(mountArgs, cephfsVolume.id)
+	mountArgs = append(mountArgs, "--client_reconnect_stale")
+	mountArgs = append(mountArgs, "--client-quota")
```
[the remark this answers is not in the document]
Comment (choury) [reply pkg/volume/cephfs/cephfs.go]: May be generate a ceph conf file for ceph-fuse? the unknown flag will be ignore then.

File: pkg/volume/cephfs/cephfs.go (hunk 6 above)
Comment (choury) [reply pkg/volume/cephfs/cephfs.go]: ceph use `asok` as socket file extension, I don't know why. But I think we can obey it.
http://docs.ceph.com/docs/master/rados/operations/monitoring/#using-the-admin-socket

File: pkg/volume/cephfs/cephfs.go (hunk 5 above)
Comment (choury) [reply pkg/volume/cephfs/cephfs.go]: Will change it soon.

File: pkg/volume/cephfs/cephfs.go (hunk 1 above)
Comment (choury) [reply pkg/volume/cephfs/cephfs.go]: Or just use `ceph-fuse` directly here? for the case `ceph-fuse` is not installed standardly.

File: pkg/volume/cephfs/cephfs.go (hunk 6 above)
Comment (cofyc) [reply pkg/volume/cephfs/cephfs.go]: didn't know that, sure.

File: pkg/volume/cephfs/cephfs.go (hunk 7 above)
Comment (choury) [reply pkg/volume/cephfs/cephfs.go]: The keyring file will be deleted after mount, whenever it succeed or failed.

File: pkg/volume/cephfs/cephfs.go (hunk 1 above)
Comment (cofyc) [reply pkg/volume/cephfs/cephfs.go]: I'm fine with either way. cc @rootfs @zhangxiaoyu-zidif

File: pkg/volume/cephfs/cephfs.go [hunk 9]
```diff
@@ -407,6 +411,12 @@ func (cephfsVolume *cephfs) execFuseMount(mountpoint string) error {
...
+	mountArgs = append(mountArgs, "--admin_socket")
+	mountArgs = append(mountArgs, filepath.Join(configPath, fmt.Sprintf("ceph-fuse.%s.asok", cephfsVolume.id)))
+	mountArgs = append(mountArgs, "--log_file")
+	mountArgs = append(mountArgs, filepath.Join(configPath, fmt.Sprintf("ceph-client.%s.log", cephfsVolume.id)))
```
[the remark this answers is not in the document]
Comment (choury) [reply pkg/volume/cephfs/cephfs.go]: @cofyc  That's too awkward, the maximum length of sock file on linux is 107, the length can exceed it if I add volume name in it.

**Comment by cofyc** [comment]
Thinking more about this. Perhaps it's not a good idea to move ceph-fuse log files to `/var/lib/kubelet`, because for log files, it's very important to logroate/compress/cleanup old files based on some policy. By default, `logrotate` package is installed on linux distros to do these tasks and ceph-common package will install a configuration file `/etc/logrotate.d/ceph-common` for log files under `/var/log/ceph`. 

So, my suggestion to avoid overriding log/sock files is to use unique string of cephfs volume in current paths:

For example:

```
/var/log/ceph/ceph-fuse-<unique-string-of-cephf-volume>.log
/var/run/ceph/ceph-fuse-<unique-string-of-cephf-volume (or hash of it)>.asok
```

To work around sock file length limit, we can use hash of <unique-string-of-cephf-volume> instead.

IIUC, `[GetVolumeName(spec *volume.Spec)](https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/cephfs/cephfs.go#L61)` should return unique string for given volume (e.g. [rbd](https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/rbd/rbd.go#L90)), but current implement returns monitor address only. Fix this first, then use it to get cephfs volume unique string? cc @rootfs

File: pkg/volume/cephfs/cephfs.go [hunk 10]
```diff
@@ -397,12 +377,43 @@ func (cephfsVolume *cephfs) execFuseMount(mountpoint string) error {
 	}
 	src += hosts[i]
 
+	configArgs := []string{}
```
Comment (batrick) [inline pkg/volume/cephfs/cephfs.go]: Why are all these config options being set? Doesn't seem relevant to stated goal of the PR.

File: pkg/volume/cephfs/cephfs.go [hunk 11]
```diff
@@ -397,12 +377,43 @@ func (cephfsVolume *cephfs) execFuseMount(mountpoint string) error {
...
+	configArgs = append(configArgs, fmt.Sprintf("[client.%s]", cephfsVolume.id))
+	configArgs = append(configArgs, fmt.Sprintf("mon_host = %s", src))
+	configArgs = append(configArgs, fmt.Sprintf("keyring = %s", keyring_file))
+	configArgs = append(configArgs, "client_reconnect_stale = true")
```
Comment (batrick) [inline pkg/volume/cephfs/cephfs.go]: This is not usually desirable. If the mount becomes stale and gets evicted, none of its in-flight I/O is valid or consistent with the other clients. I would assume the pod should be restarted and the application restarted.

Comment (choury) [reply pkg/volume/cephfs/cephfs.go]: Once we meet some network issue or mds restarted, all client can be stale. And if not reconnect it, `umont -f` is needed.

Comment (batrick) [reply pkg/volume/cephfs/cephfs.go]: > Once we meet some network issue or mds restarted, all client can be stale. And if not reconnect it, umont -f is needed.

Client mounts do not usually get disconnected by the MDS restarting. A client side network issue may result in some mounts failing but even that should not generally cause all clients to become disconnected.

> We often run a task with 30000+ container mounted cephfs, it is nightmare to handle this manually.

I would suggest that restarting the pods should be automated.

In any case, if you are adamant about this config option to solve your problem , I would ask that it remain off-by-default and that it be turned on explicitly somehow.

Comment (choury) [reply pkg/volume/cephfs/cephfs.go]: Will split it to another pr.
